### PR TITLE
[v6r17] Ignore stalled jobs for a particular site.

### PR DESCRIPTION
**Pull request by CinziaLu**
We want the StalledJobAgent to ignore stalled jobs for a particular site. Needed for BOINC sites.

### CODE REVIEW

File: WorkloadManagementSystem/Agent/StalledJobAgent.py [hunk 1]
```diff
@@ -44,6 +44,7 @@ def initialize( self ):
     self.jobDB = JobDB()
     self.logDB = JobLoggingDB()
     self.am_setOption( 'PollingTime', 60 * 60 )
+    self.ExcludeJobsfromSite = self.am_getOption( 'ExcludeJobsfromSite' )
```
Comment (andresailer) [inline WorkloadManagementSystem/Agent/StalledJobAgent.py]: At least this should be a list of sites.

File: WorkloadManagementSystem/Agent/StalledJobAgent.py [hunk 2]
```diff
@@ -115,13 +116,15 @@ def __markStalledJobs( self, stalledTime ):
...
       result = self.__getStalledJob( job, stalledTime )
-      if result['OK']:
-        self.log.verbose( 'Updating status to Stalled for job %s' % ( job ) )
-        self.__updateJobStatus( job, 'Stalled' )
-        stalledCounter += 1
-      else:
-        self.log.verbose( result['Message'] )
-        runningCounter += 1
+      site = self.jobDB.getJobAttribute( job, 'site' )
+      if self.ExcludeJobsfromSite != site:
```
Comment (andresailer) [inline WorkloadManagementSystem/Agent/StalledJobAgent.py]: This should then be 
``` python
if site in self.excludeJobsFromSite:
  continue
```
No change (indentation) in other lines necessary.

File: WorkloadManagementSystem/Agent/StalledJobAgent.py (hunk 1 above)
Comment (CinziaLu) [inline WorkloadManagementSystem/Agent/StalledJobAgent.py]: Fine.

File: WorkloadManagementSystem/Agent/StalledJobAgent.py (hunk 2 above)
Comment (CinziaLu) [inline WorkloadManagementSystem/Agent/StalledJobAgent.py]: I'll do the change.

File: WorkloadManagementSystem/Agent/StalledJobAgent.py [hunk 3]
```diff
@@ -115,6 +117,10 @@ def __markStalledJobs( self, stalledTime ):
 # jobs = jobs[:10] #for debugging
     for job in jobs:
       result = self.__getStalledJob( job, stalledTime )
+      site = self.jobDB.getJobAttribute( job, 'site' )['Value']
```
Comment (fstagni) [inline WorkloadManagementSystem/Agent/StalledJobAgent.py]: Please, few things:
- make this check not here, but in function `__updateJobStatus`, so it's less heavy
- change to:
```
siteRes = self.jobDB.getJobAttribute( job, 'site' )
if not siteRes['OK']:
  return siteRes
site = siteRes
```

**Comment by atsareg** [comment]
I do not think that Stalled jobs at some site should be simply ignored. I can imagine that for some sites the time period for which jobs are allowed to stay in the Stalled state can be longer, but not completely ignoring those.

The name for the option is not good, names starting with a verb are used for methods/functions only. For example, excludeJobsfromSite -> StalledJobsTolerantSites

**Comment by fstagni** [comment]
The remark is correct. Maybe we can say that we can increase the tolerance for stalled jobs at specific sites, like giving them more time to recover?

**Comment by CinziaLu** [comment]
It is fine for me, I will change the code.
CheersCinzia


Sent from my Samsung Galaxy smartphone.-------- Original message --------From: fstagni <notifications@github.com> Date: 04/03/2017  15:26  (GMT+01:00) To: DIRACGrid/DIRAC <DIRAC@noreply.github.com> Cc: CinziaLu <cinzia.luzzi@cern.ch>, Author <author@noreply.github.com> Subject: Re: [DIRACGrid/DIRAC] [v6r17] Ignore stalled jobs for a particular
  site. (#3270) 
The remark is correct. Maybe we can say that we can increase the tolerance for stalled jobs at specific sites, like giving them more time to recover?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/DIRACGrid/DIRAC","title":"DIRACGrid/DIRAC","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/DIRACGrid/DIRAC"}},"updates":{"snippets":[{"icon":"PERSON","message":"@fstagni in #3270: The remark is correct. Maybe we can say that we can increase the tolerance for stalled jobs at specific sites, like giving them more time to recover?"}],"action":{"name":"View Pull Request","url":"https://github.com/DIRACGrid/DIRAC/pull/3270#issuecomment-284154596"}}}

**Comment by fstagni** [comment]
Cinzia, just a reminder.

File: docs/source/AdministratorGuide/Configuration/ConfReference/Systems/WorkloadManagement/Agents/StalledJobAgent/index.rst [hunk 4]
```diff
@@ -9,30 +9,34 @@ and can be changed in the Systems/WorkloadManagement/<Instance>/JobWrapper secti
...
++----------------------------+------------------------------------------+---------------------------------+
+| *RescheduledTime*          | Age in seconds until rescheduled jobs    | RescheduledTime = 600           |
+|                            | are rescheduled                          |                                 |
+|                            |                                          |                                 |
++----------------------------+------------------------------------------+---------------------------------+
+| *CompletedTime*            | Age in seconds until completed jobs      | CompletedTime = 86400           |
+|                            | are declared failed, unless their minor  |                                 |
+|                            | status is "Pending Requests"             |                                 |
++----------------------------+------------------------------------------+---------------------------------+
+|                            | List of site for which the               | excludeJobsfromSite =           |
```
Comment (fstagni) [inline docs/source/AdministratorGuide/Configuration/ConfReference/Systems/WorkloadManagement/Agents/StalledJobAgent/index.rst]: In the third column you wrote
```
excludeJobsfromSite =
```

File: docs/source/AdministratorGuide/Configuration/ConfReference/Systems/WorkloadManagement/Agents/StalledJobAgent/index.rst [hunk 5]
```diff
@@ -9,30 +9,34 @@ and can be changed in the Systems/WorkloadManagement/<Instance>/JobWrapper secti
...
+| *StalledJobsTolerantSites* | StalledJobAgent will increase the        | siteA.cern.ch, siteB.cern.ch    |
+|                            | tolerance for stalled jobs               |                                 |
++----------------------------+------------------------------------------+---------------------------------+
+|                            | Time to be added to the StalledTimeHours |                                 |
```
Comment (fstagni) [inline docs/source/AdministratorGuide/Configuration/ConfReference/Systems/WorkloadManagement/Agents/StalledJobAgent/index.rst]: Hours, right?

**Comment by fstagni** [comment]
Still [WIP] ?

**Pull request merged**